### PR TITLE
Recordset data

### PR DESCRIPTION
**Pull request by Jay07GIT**
Fixes #.

Changes in this pull request:
-Create a new table and insert recordset_data(rsdump)

### CODE REVIEW

File: modules/api/src/main/scala/vinyldns/api/config/LimitsConfig.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/*
```
[the remark this answers is not in the document]
Comment (Jay07GIT) [reply modules/api/src/main/scala/vinyldns/api/config/LimitsConfig.scala]: Removed this commit.

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetDataRepository.scala [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+      case RecordType.PTR => "PTR"
+      case RecordType.SPF => "SPF"
+      case RecordType.SRV => "SRV"
+      case RecordType.SSHFP => "SSHFP"
+      case RecordType.TXT => "TXT"
+      case RecordType.SOA => "SOA"
+      case RecordType.DS => "DS"
+      case RecordType.NAPTR => "NAPTR"
+      case RecordType.UNKNOWN => "UNKNOWN"
+    }
```
[the remark this answers is not in the document]
Comment (Jay07GIT) [reply modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetDataRepository.scala:220]: As suggested, `Recordtype` match has been removed.

File: modules/mysql/src/main/resources/db/migration/V3.24__RecordSetData.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+USE ${dbName};
+
+CREATE TABLE recordset_data (
+  recordset_id VARCHAR(36) NOT NULL,
+  zone_id VARCHAR(36) NOT NULL,
+  fqdn VARCHAR(255) NOT NULL,
+  reverse_fqdn VARCHAR(255) NOT NULL,
+  type VARCHAR(25) NOT NULL,
+  record_data VARCHAR(4096) NOT NULL,
+  ip VARBINARY(64),
```
[the remark this answers is not in the document]
Comment (Jay07GIT) [reply modules/mysql/src/main/resources/db/migration/V3.24__RecordSetData.sql:12]: Modified as suggested.

File: modules/mysql/src/main/resources/db/migration/V3.24__RecordSetData.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  INDEX zone_id_type_fdqn_ip_rsid_rfdqn_index (zone_id, type, fqdn, ip, recordset_id, reverse_fqdn),
```
[the remark this answers is not in the document]
Comment (Jay07GIT) [reply modules/mysql/src/main/resources/db/migration/V3.24__RecordSetData.sql:13]: Given separate name for all the index.

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetRepository.scala [hunk 5]
```diff
@@ -134,7 +137,8 @@ class MySqlRecordSetRepository extends RecordSetRepository with Monitored {
             fromRecordType(i.recordSet.typ),
             toPB(i.recordSet).toByteArray,
             toFQDN(i.zone.name, i.recordSet.name),
-            i.recordSet.ownerGroupId
+            i.recordSet.ownerGroupId,
+            hashBytes(toPB(i.recordSet).toByteArray)
```
[the remark this answers is not in the document]
Comment (Jay07GIT) [reply modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetRepository.scala]: Added tests that verifying `Hash` in the name of `vinyldns` and created the new [PR](https://github.com/vinyldns/vinyldns/pull/1060) for record set blob hashing.

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetDataRepository.scala [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+  def parseIP(
+      ipString: String,
+      recordType: String
+  ): String = {
+    var ipAddress: String = null
+    val IPADDRESS_IPV4_PATTERN = {
+      "\\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\b"
+    }
+    val IPADDRESS_IPV6_PATTERN = {
+      "(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))"
```
[the remark this answers is not in the document]
Comment (Jay07GIT) [reply modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetDataRepository.scala:183]: These expressions are actually helping us to parse/extract `IPV4` and `IPV6` from the `recorddata`. So we require it to insert `ips` in `recordset_data` table

File: modules/api/src/main/scala/vinyldns/api/engine/ZoneSyncHandler.scala [hunk 7]
```diff
@@ -148,6 +150,9 @@ object ZoneSyncHandler extends DnsConversions with Monitored {
             val saveRecordChanges = time(s"zone.sync.saveChanges; zoneName='${zone.name}'")(
               recordChangeRepository.save(changeSet)
             )
+            val saveRecordSetDatas = time(s"zone.sync.saveRecordSetDatas; zoneName='${zone.name}'")(
+              recordSetDataRepository.save(changeSet)
+            )
             val saveRecordSets = time(s"zone.sync.saveRecordSets; zoneName='${zone.name}'")(
               recordSetRepository.apply(changeSet)
             )
```
[the remark this answers is not in the document]
Comment (Jay07GIT) [reply modules/api/src/main/scala/vinyldns/api/engine/ZoneSyncHandler.scala:158]: I will work on this changes and will create a new PR for it.

**Comment by Jay07GIT** [comment]
Updated transaction automicity.

File: modules/api/src/main/scala/vinyldns/api/repository/ApiDataAccessorProvider.scala [hunk 8]
```diff
@@ -35,9 +39,11 @@ object ApiDataAccessorProvider extends DataAccessorProvider[ApiDataAccessor] {
       groupChange,
       recordSet,
       recordChange,
+      recordSetData,
       zoneChange,
       zone,
-      batchChange
+      batchChange,
+      recordSetData
```
Comment (nspadaccino) [inline modules/api/src/main/scala/vinyldns/api/repository/ApiDataAccessorProvider.scala:46]: Should "recordSetData" be in this list twice?

File: modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetDataRepository.scala [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+            .bind(zoneId)
+            .update()
+            .apply()
+        }
+        logger.info(s"Deleted $numDeleted records from zone $zoneName (zone id: $zoneId)")
+      }.handleErrorWith { error =>
+        logger.error(s"Failed deleting records from zone $zoneName (zone id: $zoneId)", error)
+        IO.raiseError(error)
+      }
+    }
```
Comment (Aravindh-Raju) [inline modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetDataRepository.scala:112]: I think we are not using this function anywhere in the code. We did not mention it in our [RecordSetData repository](https://github.com/vinyldns/vinyldns/blob/12136a1d2b13bd57f2dc664c885a22cbb9031422/modules/core/src/main/scala/vinyldns/core/domain/record/RecordSetDataRepository.scala) like it's done in RecordSet repository [here](https://github.com/vinyldns/vinyldns/blob/d73cc6dd53e246357e3ae5555982d2c763553447/modules/core/src/main/scala/vinyldns/core/domain/record/RecordSetRepository.scala#L51). And we are not defining it in ZoneChangeHandler [here](https://github.com/vinyldns/vinyldns/blob/d73cc6dd53e246357e3ae5555982d2c763553447/modules/api/src/main/scala/vinyldns/api/engine/ZoneChangeHandler.scala#L39-#L40), like we did for `RecordSet.deleteRecordSetsInZone`. 

I don’t think we need this function here, as when running locally, I found that the recordset data for that particular record gets deleted automatically from `recordset_data` table when we delete the recordset. So, Can we remove this function from here ? Or does it have any use case?

Comment (Jay07GIT) [reply modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlRecordSetDataRepository.scala]: Updated `zonesynchandler.scala` for delete recordsetdata by zone id.

File: modules/api/src/main/scala/vinyldns/api/engine/ZoneChangeHandler.scala [hunk 10]
```diff
@@ -38,6 +42,9 @@ object ZoneChangeHandler {
         case Right(_) if zoneChange.changeType == ZoneChangeType.Delete =>
           recordSetRepository
             .deleteRecordSetsInZone(zoneChange.zone.id, zoneChange.zone.name)
+          executeWithinTransaction { db: DB =>
+          recordSetDataRepository
+            .deleteRecordSetDatasInZone(db,zoneChange.zone.id, zoneChange.zone.name)}
             .attempt
             .flatMap { _ =>
               zoneChangeRepository.save(zoneChange.copy(status = ZoneChangeStatus.Synced))
```
[the remark this answers is not in the document]
Comment (Jay07GIT) [reply modules/api/src/main/scala/vinyldns/api/engine/ZoneChangeHandler.scala]: Wrapped `deleteRecordSetDatasInZone` into `for yield` and recommended .

**Review by remerle — APPROVED** [review]

**Pull request merged**